pylint/checkers/typecheck.py

```diff
@@ -631,7 +631,7 @@ def _no_context_variadic_keywords(node, scope):
 
 def _no_context_variadic_positional(node, scope):
     variadics = ()
-    if isinstance(scope, nodes.Lambda) and not isinstance(scope, nodes.FunctionDef):
+    if isinstance(scope, nodes.Lambda):
```

Comment:
This seem to fix the issue from #5416 without causing regression, but I don't understand the else branch. With this change the else is no longer covered.

Reply:
Seeing the removed conditional it was for treating ``FunctionDef`` (duh). But it seems it could be treated generically. All tests works well with this change. The really hard question is how to trigger this branch now and is it possible at all ? Did you try running the primer locally with code coverage ? That could help find a covering functional test. (The online code coverage is calculated on standard tests only).

Reply:
https://github.com/PyCQA/pylint/blob/e33bc6377bbbf6135888cdadecdfc3baa8497a24/pylint/checkers/typecheck.py#L636-L642

I have looked at the `else` for a bit and in the current tests there is no test where `node != call` and none where `variadics == ()`. Therefore this whole `else` block seems meaningless as doing `variadics = node.starargs + node.kwargs` for all `node`'s works as well. We might be okay by removing both the `if` and the `else`.

The commit that added this was https://github.com/PyCQA/pylint/commit/99ce3cdfd3bc1bf2891fd807f97ee3cd1ec26c21. It mentions a lack of context recognition, that might have been solved now. This commit fixes https://github.com/PyCQA/pylint/issues/2918. @Pierre-Sassoulas do you know anything about this?

The commit prior to that also makes little sense https://github.com/PyCQA/pylint/commit/d95eb088442fad414772b85873c08ac94587ad9a. In `_no_context_variadic_keywords` there is a check for `scope` not being `FunctionDef`. However, the only call to `_no_context_variadic_keywords` is guarded by a check that `scope` is `FunctionDef` on L562.

I'm not really sure what this code is trying to achieve. Based on the expansion of the if statement on L638-640 on the current code compared to the original I feel like this might have been causing false positives and the list of accepted `statement` types was expanded to resolve those without looking at whether the method actually made sense. See for example this issue https://github.com/PyCQA/pylint/issues/3563.

Reply:
Thanks for feedback here and sorry for being late replying. I tested a bit more and with that first version of the patch ( for reference it was https://github.com/PyCQA/pylint/pull/5417/commits/4138aa61442f3f636cf183a23c382eb9c43962b6 ) both with unit tests or primer tests the `else:` branch was never used.

I can also confirming that with a simplified version like above all tests are also passing, I think @DanielNoord this is what you suggested.

```python
def _no_context_variadic_positional(node, scope):
    variadics = node.starargs + node.kwargs
    return _no_context_variadic(node, scope.args.vararg, nodes.Starred, variadics)
```

I pushed a new version of the patch using this implementation and also added a few more tests. This test file had text for `*args` and `**kwargs`, so I added the same test for `**kwargs`, even if it was not problematic, I figured it might be good to have it tested as well.

I also added changelog entry etc and I'm marking as ready for review, so if you like, maybe this can be merged. Frankly, I don't fully understand the previous code here, so if you prefer waiting until we get a better understanding of what the code was trying to achieve, that's of course OK with me, but I don't think I'll be able to help with understanding the previous code.

Thanks again for your suggestions

Reply:
I did some digging, the else was added in 99ce3cdfd3bc1bf2891fd807f97ee3cd1ec26c21, and the functional test added together with it was:

```python
def func(one, two, three):
    return one + two + three


CALL = lambda *args: func(*args)
```

I'll look into it in more detail tomorrow. I was merging approved MR for 2.13 and this is the last one that is approved and also passing tests :)